Fee cutoff

djseeds opened the pull request:
Implementation of #278 

Add fee-based cut-off during route construction

Extend SendRequest proto to allow users to specify one of:
max_fee_satoshis: "I will pay at most x satoshi in fees"
max_fee_ratio: "I will pay at most x * amt in fees".

rpcserver: Modify SendPayment to accept MaxFee parameter.
router: Modify newRoute function to abandon routes with fees greater than maxFee.
lncli: Add max_fee_satoshis and max_fee_ratio options to sendpayment command.

djseeds commented on rpcserver.go:
```diff
@@ -1530,6 +1530,20 @@ func (r *rpcServer) SendPayment(paymentStream lnrpc.Lightning_SendPaymentServer)
+				case *lnrpc.SendRequest_LimitSatoshis:
+					*maxFee = btcutil.Amount(nextPayment.GetLimitSatoshis())
+				case *lnrpc.SendRequest_LimitRatio:
+					*maxFee = btcutil.Amount(nextPayment.GetLimitRatio()) * amt
```
Should I do any more checking here on the limit ratio (e.g. check if it is negative)? As it stands now I think it will simply fail to find a route.

djseeds commented on rpcserver.go:
```diff
@@ -1663,7 +1677,7 @@ func (r *rpcServer) SendPaymentSync(ctx context.Context,
 		Amount:      amtMSat,
 		PaymentHash: rHash,
-	})
+	}, nil)
```
Realized I should have a non-nil maxFee here too...
This means I will need to do the type switch again, so I think I should make it a function. Where would be a good place to put a utility function like that?

djseeds commented on routing/router_test.go:
```diff
@@ -127,7 +127,7 @@ func TestFindRoutesFeeSorting(t *testing.T) {
 	paymentAmt := lnwire.NewMSatFromSatoshis(100)
 	target := ctx.aliases["luoji"]
-	routes, err := ctx.router.FindRoutes(target, paymentAmt)
+	routes, err := ctx.router.FindRoutes(target, paymentAmt, nil)
```
I will need to add some tests with non-nil maxFee

djseeds replied:
Added tests. I believe integration tests of the RPC are still needed.

djseeds commented on routing/pathfind.go:
```diff
@@ -196,13 +196,14 @@ func (s sortableRoutes) Swap(i, j int) {
 // the source to the target node of the path finding attempt.
 func newRoute(amtToSend lnwire.MilliSatoshi, pathEdges []*ChannelHop,
-	currentHeight uint32) (*Route, error) {
+	currentHeight uint32, maxFee *btcutil.Amount) (*Route, error) {
```
I was hoping to use the interface created by the `oneof` as a parameter, but it is not exported. Instead, I take a pointer to the maximum fee in satoshis, or `nil` if there is no max fee. Let me know if there is a more idiomatic way to have an optional argument in Go, or if I'm totally missing another approach

djseeds commented on lnrpc/rpc.proto:
```diff
@@ -461,6 +461,16 @@ message SendRequest {
+        // Abandon the route if fees greater than limit_satoshis
+        uint64 limit_satoshis = 8 [json_name = "limit_satoshis"];
+        //Abandon the route if fees greater than limit_ratio * amt
+        double limit_ratio = 9 [json_name = "limit_ratio"];
```
The issue specifies a limit percentage, but I thought it might make more sense to specify it as a ratio. Let me know if I should change it to be a percentage.

djseeds commented on lnrpc/rpc.proto:
```diff
@@ -461,6 +461,16 @@ message SendRequest {
+        // Abandon the route if fees greater than amt
+        bool absolute_cut_off = 7 [json_name = "absolute_cut_off"];
+        // Abandon the route if fees greater than limit_satoshis
+        uint64 limit_satoshis = 8 [json_name = "limit_satoshis"];
```
Now that I know that "MilliSatoshi are the native unit of the Lightning Network", should this be changed to MilliSatoshis?

jimpo commented on routing/router.go:
```diff
@@ -967,13 +968,15 @@ type LightningPayment struct {
 // within the network to reach the destination. Additionally, the payment
 // preimage will also be returned.
-func (r *ChannelRouter) SendPayment(payment *LightningPayment) ([32]byte, *Route, error) {
+func (r *ChannelRouter) SendPayment(payment *LightningPayment, maxFee *lnwire.MilliSatoshi) ([32]byte, *Route, error) {
```
Seems like max fee should maybe be a field on `LightningPayment` instead.

djseeds replied:
@jimpo I totally agree, I'll make those changes now

jimpo commented on lnrpc/rpc.proto:
```diff
@@ -461,6 +461,14 @@ message SendRequest {
     string payment_request = 6;
+
+    // Cut off specifying a fee value above which path finding will be short circuited
+    oneof fee_cut_off {
```
I could imagine a user wanting to specify both.

djseeds replied:
Interesting, can I have a for instance?

jimpo replied:
I just don't see the options as mutually exclusive. A user might want to say "I will pay up to 20% of the transaction value in fees with a hard limit of 0.1 BTC."

It wouldn't complicate the implementation much either. You just set max `fee = min(absolute cutoff, ratio cutoff * amount)`.

halseth replied:
I would argue that this logic could be moved to the caller, such that we can have only the `limit_satoshis` field, and make the caller do the necessary calculation. That would simplify the code a bit?

djseeds replied:
Yeah, that could be the way to go. I don't think the code is overly complex as it is, but it would simplify it a bit. Would certainly simplify RPC testing though.

djseeds replied:
Should this limit be specified in millisatoshis? The issue says satoshis, but I think that removes a ton of granularity between 0 and 1 satoshi. Default fee is 1 millisatoshi, correct?

djseeds replied:
@halseth I am attempting to make this a singular field as you described, but it seems that proto3 provides no way to distinguish between unset fields and fields that were explicitly set to 0 (i.e. unset fields are set to their 0 value). With the oneof field, it was simple to first check against `nil`. Any guidance on how to overcome this? Should we require the caller to always set this value if they would like routes with fee > 0? 

On second thought, I guess that would be a useful function, so we would look for fee-less routes by default, and if none exist then the user could set their real max fee. It seems more intuitive for unset to mean "don't care", rather than 0, though.

Let me know what you think!

halseth replied:
I was thinking maybe require the caller to set its `maxfee`. The only downside I see by doing this, is that it might break apps that are already using this API (since maxFee would default to 0 and they would struggle to find 0 fee routes). But I would argue that's a change you can do in an alpha release 😄 

Also remember that you must update `lncli` (found in `cmd/`) to accept the new parameters, and use the new gRPC fields.

halseth commented on routing/pathfind.go:
```diff
@@ -196,13 +196,14 @@ func (s sortableRoutes) Swap(i, j int) {
 // the source to the target node of the path finding attempt.
 func newRoute(amtToSend lnwire.MilliSatoshi, pathEdges []*ChannelHop,
-	currentHeight uint32) (*Route, error) {
+	currentHeight uint32, maxFee *lnwire.MilliSatoshi) (*Route, error) {
```
To avoid using a nilable parameter, you could make it non-nilable, and then when you extract the `maxFee` you set `maxFee = payAmt` if no user specified max :)

djseeds replied:
I thought about that, but I wasn't sure if it would be right to set a hard limit at `amt`. Would we want it to be impossible for the user to pay a large fee if they really need/want to? E.g. Alice wants to send 1 satoshi to Carol ASAP, but the only route between them requires a 2 satoshi fee. Should Alice never be able to make such a payment?

halseth replied:
Good point! What about setting it to some very large number in that case instead?

djseeds replied:
Sounds good! How about `math.MaxInt64`?

halseth replied:
Maybe you could use the constant `rpcserver.maxPaymentMSat`. Cannot imagine anybody want to pay a higher fee than the max allowed payment amount :P

djseeds replied:
Perfect, thanks!

djseeds replied:
@halseth I ended up bringing back the nilable parameter. The reason for this is that roasbeef requested that I make the MaxFee field of the LightningPayment struct optional. Besides at extraction, this function seems to be the most convenient place to check for `nil` on that field. I could certainly move that functionality up to router.SendPayment, but I believe the outcome would be the same. Let me know what you think!

djseeds commented:
I think this should be ready for review now. Let me know what you think!

Roasbeef commented on cmd/lncli/commands.go:
```diff
@@ -766,6 +766,10 @@ var sendPaymentCommand = cli.Command{
 		},
+		cli.Int64Flag{
+			Name:  "max_fee",
+			Usage: "Maximum fee to pay in millisatoshis.",
```
It's a bit odd for this unit to be in `mSAT` when currently we accept/display `SAT` everywhere uniformly within the RPC interface (other than when modifying the current fee schedule). This was a purposeful decision, as atm, I don't see users utilizing the `mSAT` payment granularity at this stage.

djseeds replied:
Thanks for the feedback! I originally had this in satoshis and had commented on this PR "Now that I know that "MilliSatoshi are the native unit of the Lightning Network", should this be changed to MilliSatoshis?" I think my squash removed my comments and questions from the actual code but they should be visible on the PR page.

After changing it internally to use msat, I spoke to you about the fee limit on #dev-help and you said "fee limit should operate on msat, as fees are paid in msat".

Shall I change it to operate internally on msat level, but externally on sat level?

Roasbeef commented on lnd_test.go:
```diff
@@ -1029,6 +1029,7 @@ func testSingleHopInvoice(net *networkHarness, t *harnessTest) {
 		PaymentHash: invoiceResp.RHash,
 		Dest:        net.Bob.PubKey[:],
 		Amt:         paymentAmt,
+		MaxFee:      int64(lnwire.MaxPaymentMSat),
```
For the sake of minimizing the diff: I'm not sure it's necessary to actually specify this value. If it is (due to the way the feature has been implemented), then it should be amended, as we shouldn't _force_ users to always specify a `MaxFee` parameter. Instead, it should strictly be optional.

djseeds replied:
Originally this value was a pointer which could be left `nil` for no fee limit. After this conversation: https://github.com/lightningnetwork/lnd/pull/320#discussion_r139392829 , this change was made because it seems that proto3 provides no way to distinguish between unset singular fields and fields that were explicitly set to 0. This would be resolved by bringing back the `oneof` field.

Roasbeef commented on lnd_test.go:
```diff
@@ -1464,6 +1468,150 @@ func testMultiHopPayments(net *networkHarness, t *harnessTest) {
 	}
 }
 
+func testMaxFee(net *networkHarness, t *harnessTest) {
```
Missing a `godoc` style comment describing what functionality is under test by this function.

Roasbeef commented on lnd_test.go:
```diff
@@ -1464,6 +1468,150 @@ func testMultiHopPayments(net *networkHarness, t *harnessTest) {
+		t.Fatalf("carol didn't see the alice->bob channel before timeout: %v", err)
+	}
+
+	// Using Carol as the source, pay to the 5 invoices from Bob created above.
```
Comment doesn't match the code itself. Looks like it's been copy-pasta'd from elsewhere.

Roasbeef commented on lnd_test.go:
```diff
@@ -1464,6 +1468,150 @@ func testMultiHopPayments(net *networkHarness, t *harnessTest) {
+		t.Fatalf("unable to create payment stream for carol: %v", err)
+	}
+
+	// Concurrently pay off all 5 of Bob's invoices. Each of the goroutines
```
Comment doesn't match the code itself. Looks like it's been copy-pasta'd from elsewhere.

Roasbeef commented on lnd_test.go:
```diff
@@ -1464,6 +1468,150 @@ func testMultiHopPayments(net *networkHarness, t *harnessTest) {
+
+		if resp, err := carolPayStream.Recv(); err != nil {
+			t.Fatalf("payment stream has been closed: %v", err)
+		} else if resp.PaymentError != "" {
```
This can be re-written a bit to make the flow of the test a bit more explicit: 
```go
if resp, err := carolPayStream.Recv(); err != nil {
    t.Fatalf("payment stream has been closed: %v", err)
}

switch {
    case resp.PaymentError != "" && test.expecteSucess:
        // fail the test
    case resp.PaymentError == "" && !test.expectSucess:
       // fail the test
}
```

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -461,6 +461,10 @@ message SendRequest {
+
+    // Maximum fee to pay in millisatoshis. Routes with fees greater
+    // than max_fee will be abandoned during path finding.
+    int64 max_fee = 7 [json_name = "max_fee"];
```
The initial issue specifies the ability to configure either: a max fee (absolute), or a ratio. Is there a precise reason you opted to only go with the first option?

djseeds replied:
In my original PR I had a `oneof` field which allowed the user to specify all three of the options suggested in the issue. I am realizing I should not have squashed this issue down to one commit, so you could have more context.

I went with the singular field after a conversation with halseth where he argued "that this logic could be moved to the caller, such that we can have only the limit_satoshis field, and make the caller do the necessary calculation." You can see that conversation here:
https://github.com/lightningnetwork/lnd/pull/320#discussion_r139392829

I can certainly bring back the `oneof` if you think that would be best.

Roasbeef commented on routing/router_test.go:
```diff
@@ -146,6 +146,82 @@ func TestFindRoutesFeeSorting(t *testing.T) {
+	paymentAmt := lnwire.NewMSatFromSatoshis(100)
+	target := ctx.aliases["luoji"]
+
+	// There are two possible paths from roasbeef to luo ji, with fees of
```
Missing a period at the end of sentences in a few comments within this test case (won't comment on them all in order to reduce review comment clutter).

Roasbeef commented:
Hi @djseeds apologies for the delay with the final review of this PR. Can you rebase this to the latest master? Thanks!

djseeds commented:
@Roasbeef No problem, I know you guys are busy working on some great things and I'm just glad to help! 

I've rebased to current master, and I think I have addressed all of your comments. Before this can be merged, though, I think I need some guidance on how the fee cutoff should interact with the route cache in the [router](https://github.com/lightningnetwork/lnd/blob/f20cb89982e0d9a66cf797383693241bf6c4acb5/routing/router.go#L951).

Currently the fee cutoff does not filter cached routes, so it is possible for a payment to be sent with fee > maxFee if there has been a previous payment sent to the same node. This is also the reason my tests are currently failing. I've described the issue a bit further in #278 .

Roasbeef commented:
Replaced by #714.
